standard/statements.md

```diff
@@ -1112,33 +1112,50 @@ If the *foreach_statement* contains both or neither `ref` and `readonly`, the it
-  - Member lookup is performed on `E` with the identifier `MoveNext` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a method group, an error is produced and no further steps are taken.
-  - Overload resolution is performed on the method group with an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, or its return type is not `bool`, an error is produced, and no further steps are taken.
-  - The collection type is `X`, the enumerator type is `E`, and the iteration type is the type of the `Current` property. The `Current` property may include the `ref` modifier, in which case, the expression returned is a *variable_reference* ([§9.5](variables.md#95-variable-references)) that is optionally read-only.
-- Otherwise, check for an enumerable interface:
-  - If among all the types `Tᵢ` for which there is an implicit conversion from `X` to `IEnumerable<Tᵢ>`, there is a unique type `T` such that `T` is not `dynamic` and for all the other `Tᵢ` there is an implicit conversion from `IEnumerable<T>` to `IEnumerable<Tᵢ>`, then the collection type is the interface `IEnumerable<T>`, the enumerator type is the interface `IEnumerator<T>`, and the iteration type is `T`.
-  - Otherwise, if there is more than one such type `T`, then an error is produced and no further steps are taken.
-  - Otherwise, if there is an implicit conversion from `X` to the `System.Collections.IEnumerable` interface, then the collection type is this interface, the enumerator type is the interface `System.Collections.IEnumerator`, and the iteration type is `object`.
-  - Otherwise, an error is produced, and no further steps are taken.
+- «GetEnumerator» is a `GetEnumerator` method.
+- «MoveNext» is a `MoveNext` method with a `bool` return type.
+- «IEnumerable»\<T> is the `System.Collections.Generic.IEnumerable<T>` interface.
+- «IEnumerator»\<T> is the `System.Collections.Generic.IEnumerator<T>` interface.
+
+In addition, the following modifications are made to the steps in §13.9.5.1:
+
+Before the process described in §13.9.5.1, the following steps are taken:
 
-The above steps, if successful, unambiguously produce a collection type `C`, enumerator type `E` and iteration type `T`, `ref T`, or `ref readonly T`. A `foreach` statement of the form
+- If the type `X` of *expression* is an array type then there is an implicit reference conversion from `X` to the `IEnumerable` interface (since `System.Array` implements this interface). The collection type is the `IEnumerable` interface, the enumerator type is the `IEnumerator` interface and the iteration type is the element type of the array type `X`. (sync)
```

Comment:
In 17.2.3 we already say:

> A single-dimensional array `T[]` implements the interface `System.Collections.Generic.IList<T>` and its base interfaces.

So can't we say the collection type is `IEnumerable<A>` and the interface type is `IEnumerator<A>` where `A` is the element type of the array type `X`?